File: leetcode/maxNumberOfBalloons.js
/**
 * @param {string} text
 * @param {string} subStr
 * @return {number}
 */
var strCount = function (text, subStr) {
  if (text.match(new RegExp(subStr, "g")) != null) {
    return text.match(new RegExp(subStr, "g")).length;
  }
  return 0;
};

/**
 * @param {string} text
 * @return {number}
 */
var maxNumberOfBalloons = function (text) {
  let n = strCount(text, "b");
  while (n > 0) {
    if (n == 0) {
      break;
    }
    if (strCount(text, "a") >= n) {
      if (strCount(text, "n") >= n) {
        if (strCount(text, "o") >= n * 2) {
          if (strCount(text, "l") >= n * 2) {
            break;
          } else {
            n -= 1;
          }
        } else {
          n -= 1;
        }
      } else {
        n -= 1;
      }
    } else {
      n -= 1;
    }
  }
  return n;
};

/**
 * @param {string} text
 * @return {number}
 */
var maxNumberOfBalloons2 = function (text) {
  return Math.min(
    Math.floor(strCount(text, "b")),
    Math.floor(strCount(text, "a")),
    Math.floor(strCount(text, "l") / 2),
    Math.floor(strCount(text, "o") / 2),
    Math.floor(strCount(text, "n"))
  );
};

console.log(maxNumberOfBalloons2("baxxllonob"));
console.log(maxNumberOfBalloons2("loboonbalxbaallpoooonbb"));
console.log(maxNumberOfBalloons2("leetcode"));
console.log(maxNumberOfBalloons2("balon"));
console.log(maxNumberOfBalloons(""));

File: leetcode/maxNumberOfBalloons.py
class Solution:
    def maxNumberOfBalloons(self, text):
        n = text.count('b')
        while n > 0:
            if n == 0:
                break
            if text.count('a') >= n:
                if text.count('n') >= n:
                    if text.count('o') >= n * 2:
                        if text.count('l') >= n * 2:
                            break
                        else:
                            n -= 1
                    else:
                        n -= 1
                else:
                    n -= 1
            else:
                n -= 1
        return n


answer = Solution()
print(answer.maxNumberOfBalloons("nlaebolko"))
print(answer.maxNumberOfBalloons("loonbalxballpoon"))
print(answer.maxNumberOfBalloons("loboonbalxbaallpoooonbb"))
print(answer.maxNumberOfBalloons("bleetcode"))

'''
Given a string text, you want to use the characters of text to form as many instances of the word "balloon" as possible.

You can use each character in text at most once. Return the maximum number of instances that can be formed.

Example 1:

Input: text = "nlaebolko"
Output: 1
Example 2:

Input: text = "loonbalxballpoon"
Output: 2
Example 3:

Input: text = "leetcode"
Output: 0

class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        return min(text.count('b'), text.count('a'), text.count('l') // 2, text.count('o') // 2, text.count('n'))

class Solution:
    def maxNumberOfBalloons(self, t: str) -> int:
        return min(t.count(c) // int(cnt) for c, cnt in zip('balon', '11221'))

class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        return min(text.count(c) // 'balloon'.count(c) for c in 'balon')
'''